custom_model_runner/datarobot_drum/drum/typeschema_validation.py

```diff
@@ -546,9 +546,7 @@ def __init__(
         """
         self._using_default_type_schema = False
         if not type_schema and use_default_type_schema:
-            from datarobot_drum.drum.common import (
-                read_model_metadata_yaml,
-            )  # local import to prevent cyclic dependency
+            from datarobot_drum.drum.model_metadata import read_model_metadata_yaml
 
             type_schema = read_model_metadata_yaml(
```

Comment:
@eric-s-s could you recommend how to resolve this circular import?

Reply:
yes.  this won't be easy.  conceptually, `common` should not be importing _from_ other places except for the most basic stuff like `enum` and `exceptions`  so these two import lines are _extremely_ suspicious

```
from datarobot_drum.drum.model_metadata import HyperParameterTrafaret

from datarobot_drum.drum.typeschema_validation import (
    get_type_schema_yaml_validator,
    revalidate_typeschema,
)
```

You should probably keep all of the type_schema validation _only_ in that file.  

notice that all of the funcitonality that uses this is in:

```python 
def read_model_metadata_yaml(code_dir) -> PythonTypingOptional[dict]:
    code_dir = Path(code_dir)
    config_path = code_dir.joinpath(MODEL_CONFIG_FILENAME)
    if config_path.exists():
        with open(config_path) as f:
            try:
                model_config = load(f.read(), MODEL_CONFIG_SCHEMA)  #  THIS IS THE ONLY USE OF THE SCHEMA
                if "typeSchema" in model_config:
                    revalidate_typeschema(model_config["typeSchema"])
                model_config = model_config.data
            except YAMLError as e:
                if "found a blank string" in e.problem:
                    print("The model_metadata.yaml file appears to be empty.")
                else:
                    print(e)
                raise SystemExit(1)
```

## actions

1. `read_model_metadata_yaml` should be in its own place
2. the type schema validator should not be reading metadata, it should _only_ validate existing metadata and should always have that passed in.  
3. none of this should be in common

the main ones are to separate the validation logic from the reader logic.  if the validator only ever deals with already-loaded data, then you won't need to worry about this inside of the validator.  whoever calls the validator should both read the data and then pass it to it's caller and that should take care of your issues.